python_modules/libraries/dagster-dbt/dagster_dbt/core/utils.py

```diff
@@ -300,3 +311,62 @@ def get_future_completion_state_or_err(futures: List[Union[Future, Any]]) -> boo
         if exception:
             raise exception
     return True
+
+
+T = TypeVar("T")
+P = TypeVar("P")
+
+
+def imap(
+    executor: ThreadPoolExecutor,
+    iterable: Iterator[T],
+    func: Callable[[T], P],
+    block_on_enqueuing_task_completion: bool = False,
+) -> Iterator[P]:
```

Comment:
Consider using `FuturesAwareThreadPoolExecutor`: https://github.com/dagster-io/dagster/blob/a3a44b7f0cfd55c121afb6280b291315b25ffac9/python_modules/dagster/dagster/_core/utils.py#L146-L178

So that we can compound the benefits of some of this logic onto that abstraction.

Reply:
What kind of integration are you thinking? Should still work if passed a `FuturesAwareThreadPoolExecutor`, but I think regardless this impl will need to keep a separate ordered list of futures because `FuturesAwareThreadPoolExecutor` discards its completed futures

Reply:
I was thinking that we could pass some override to initialize `_tracked_futures` as something other than `WeakSet`. 

My main concern is the current implementation of `work_queue` in this PR never discards a future. Somehow, I would want the memory of that to be a bit more bounded.

Reply:
It is a little cumbersome to try to get this to work with `FuturesAwareThreadPoolExecutor` because of the need for structure/ordering in the list of futures and the ability to introspect into the tracked futures in a particular order. It feels like the utility of a class like `FuturesAwareThreadPoolExecutor` is to not have to worry about the internals of bookkeeping `Future`s & to be able to query high-level status, but in this case we need pretty direct access to the underlying data structure. For now, updated the fn to use a deque which should let us drop the `Future`s as we complete them

Reply:
Cool, that's fine w me.